cmd/serve.go

```diff
@@ -15,107 +15,316 @@
 package cmd
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
-	"html/template"
+	stdlog "log"
 	"net/http"
 	"os"
+	"os/signal"
+	"sort"
 	"strings"
+	"syscall"
+	"time"
 
 	"github.com/spf13/cobra"
 
-	"github.com/ossf/scorecard/v5/clients/githubrepo"
-	"github.com/ossf/scorecard/v5/clients/ossfuzz"
+	"github.com/ossf/scorecard/v5/checker"
+	"github.com/ossf/scorecard/v5/clients"
+	"github.com/ossf/scorecard/v5/clients/localdir"
+	pmc "github.com/ossf/scorecard/v5/cmd/internal/packagemanager"
+	docs "github.com/ossf/scorecard/v5/docs/checks"
 	"github.com/ossf/scorecard/v5/log"
 	"github.com/ossf/scorecard/v5/options"
 	"github.com/ossf/scorecard/v5/pkg/scorecard"
+	"github.com/ossf/scorecard/v5/policy"
 )
 
-// TODO(cmd): Determine if this should be exported.
+type server struct {
+	logger *log.Logger
+	opts   *options.Options
+}
+
+type scorecardRequest struct {
+	Repo        string   `json:"repo"`
+	Local       string   `json:"local,omitempty"`
+	NPM         string   `json:"npm,omitempty"`
+	PyPI        string   `json:"pypi,omitempty"`
+	RubyGems    string   `json:"rubygems,omitempty"`
+	Nuget       string   `json:"nuget,omitempty"`
+	Checks      []string `json:"checks,omitempty"`
+	Commit      string   `json:"commit,omitempty"`
+	CommitDepth int      `json:"commit_depth,omitempty"`
+	ShowDetails bool     `json:"show_details,omitempty"`
+	Format      string   `json:"format,omitempty"`
+	LogLevel    string   `json:"log_level,omitempty"`
+	Probes      []string `json:"probes,omitempty"`
+	FileMode    string   `json:"file_mode,omitempty"`
+	PolicyFile  string   `json:"policy_file,omitempty"`
+}
+
+func newServer(logger *log.Logger, opts *options.Options) *server {
+	return &server{
+		logger: logger,
+		opts:   opts,
+	}
+}
+
+func (s *server) handleScorecard(w http.ResponseWriter, r *http.Request) {
+	var req scorecardRequest
+	if r.Method == http.MethodPost {
+		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+			http.Error(w, "invalid request body", http.StatusBadRequest)
+			return
+		}
+	} else {
+		req.Repo = r.URL.Query().Get("repo")
+		req.Local = r.URL.Query().Get("local")
+		req.NPM = r.URL.Query().Get("npm")
+		req.PyPI = r.URL.Query().Get("pypi")
+		req.RubyGems = r.URL.Query().Get("rubygems")
+		req.Nuget = r.URL.Query().Get("nuget")
+		req.Checks = strings.Split(r.URL.Query().Get("checks"), ",")
+		req.Commit = r.URL.Query().Get("commit")
+		req.ShowDetails = r.URL.Query().Get("show_details") == "true"
+		req.Format = r.URL.Query().Get("format")
+		req.LogLevel = r.URL.Query().Get("log_level")
+		req.Probes = strings.Split(r.URL.Query().Get("probes"), ",")
+		req.FileMode = r.URL.Query().Get("file_mode")
+		req.PolicyFile = r.URL.Query().Get("policy_file")
+	}
+
+	// Set options
+	s.opts.Repo = req.Repo
+	s.opts.Local = req.Local
+	s.opts.NPM = req.NPM
+	s.opts.PyPI = req.PyPI
```

Comment:
Right now we have one `options.Options`, which is passed in `serveCmd`, and then this one copy is modified for each request. This seems like a race condition.

we should create a new struct for each request. either through `options.New`, or manually if we want to avoid the env var parsing.